core/embed/rust/src/trezorhal/fatal_error.rs

```diff
@@ -78,19 +87,13 @@ macro_rules! unwrap {
 macro_rules! ensure {
     ($what:expr, $error:expr) => {
         if !($what) {
-            fatal_error!(stringify!($what), $error);
+            crate::trezorhal::fatal_error::__fatal_error($error, file!(), line!());
         }
     };
 }
 
 macro_rules! fatal_error {
```

Comment:
i am somewhat unhappy about calling differently named function (or macro) in rust vs C code when i want to show error and shutdown. How about getting rid of this macro, create the same `error_shutdown` and `error_shutdown_ex` structure as was introduced in C and leave `__fatal_error` for internal purposes only (called from `ensure` and `unwrap`, while regular `error_shutdown` calls would lose the file and line info as C calls do).

Also, i think that we can move most of the code from this file (without the `shutdown` function that interfaces with C) to e.g. `rust/src/error_handling.rs`, possible move the `panic` implementation there too from lib.rs.

cc also @matejcik , what do you think?

Reply:
It makes some amount of sense to me like it is now.
Our `fatal_error!` is an equivalent to `panic!`, which has just a single string reason. rust enriches it with file+line information and passes it on. This is an "internal error" with some fancy paint on it.
Or, you want to invoke an explicitly user-facing RSOD, then you call the _function_ `error_shutdown`.
These are two distinct operations. C folds them into kiiind of one-and-the-same, which I think is a small bit less semantically correct? But not enough for me to care.

In any case, I wouldn't want to introduce "error_shutdown_ex" (the "_ex" being a specifically C thing), and I don't have any other ideas that would be less confusing. We can do `error_shutdown!("just message")` and `error_shutdown("title", "message", "footer")`, or we can extend `error_shutdown!` to accept either three or one arguments, either sounds just weird.
(other options anyone?)

> Also, i think that we can move most of the code from this file (without the shutdown function that interfaces with C) to e.g. rust/src/error_handling.rs, possible move the panic implementation there too from lib.rs.

my followup PR already moves the macros. I would keep the functions (error_shutdown, __fatal_error) here, seeing as they all transitively depend on the ffi `shutdown()` call, so they are all trezorhal-y.
the traits could be moved, but perhaps do the opposite? keep them here, move the panic impls here too (is it legal to do it that way?) and call this whole module `shutdown` or something.

Reply:
Do you mind if we do it in a separate PR or in a follow-up PR #3951 to be able to merge this soon?

Reply:
All right, i see your point, lets keep it as is then